### PR TITLE
Add Test proving a Bug in SCIM Filters.

### PR DESCRIPTION
**Pull request by sputnik27**
Related to osiam/resource-server#7

The test is based on [0]. It demonstrates the behavior as described in [1].
In this commit, the test is @Ignore'd because it fails (as expected)
until [1] is fixed. The @Ignore must be removed once the bug is fixed.

[0] https://gist.github.com/sschum/ad6a09a2e70ae95ad1ce
[1] https://github.com/osiam/resource-server/issues/7

### CODE REVIEW

File: src/test/java/org/osiam/client/BugFilterWithNot.java [hunk 1]
```diff
@@ -0,0 +1,109 @@
...
+    public void setup() {
+        accessToken = OSIAM_CONNECTOR.retrieveAccessToken(ADMIN_USER_NAME, ADMIN_USER_PASSWORD, Scope.ADMIN);
+    }
+
+    @Ignore("This test fails at the moment because the bug is not yet fixed. @Ignore must be removed after the bug-fix.")
+    @Test
+    public void filter_groups_where_the_user_is_not_included_in() throws Exception {
+        // check precondition
+        Group expectedGroup = getGroupByName(NAME_GROUP_TWO);
+        assertNotContainsUser(expectedGroup, TEST_USER_ID);
```
Comment (wallner) [inline src/test/java/org/osiam/client/BugFilterWithNot.java]: I don't think this assertion is necessary, after it is just asserting that the test data as described in the seed is actually in the DB. I would drop it.

File: src/test/java/org/osiam/client/BugFilterWithNot.java [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+        assertThat(String.format("Can not find expected group %s!", NAME_GROUP_TWO),
+                result.getResources(), hasItem(new Group.Builder(NAME_GROUP_TWO).build()));
+    }
+
+    @Ignore("This test fails at the moment because the bug is not yet fixed. @Ignore must be removed after the bug-fix.")
+    @Test
+    public void filter_users_where_not_a_member_of_group() throws Exception {
+        // check precondition
+        Group expectedGroup = getGroupByName(NAME_GROUP_TWO);
+        assertNotContainsUser(expectedGroup, TEST_USER_ID);
```
Comment (wallner) [inline src/test/java/org/osiam/client/BugFilterWithNot.java]: The same holds true here. I don't think that we have to make sure the fixture is set up correctly.

**Comment by sputnik27** [comment]
Thanks for your remarks, @wallner. You are absolutely right, and this makes the code shorter and more readable. I updated my PR.

**Comment by tkrille** [comment]
`bug-filter-with-not-issue`? Can you think of something else?

File: src/test/java/org/osiam/client/BugFilterWithNot.java [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+
+    private Group getGroupByName(String groupName) throws Exception {
+        Query query = new QueryBuilder().filter(String.format("displayName eq \"%s\"", groupName)).build();
+
+        SCIMSearchResult<Group> result = OSIAM_CONNECTOR.searchGroups(query, accessToken);
+        if (!result.getResources().isEmpty()) {
+            return result.getResources().get(0);
+        }
+
+        throw new Exception("The expected group was not found. This is an error with the test data.");
```
Comment (tkrille) [inline src/test/java/org/osiam/client/BugFilterWithNot.java]: Don't use conditions or `throw`s. The test will fail anyway if the database fixture could not be set up.

Comment (tkrille) [inline src/test/java/org/osiam/client/BugFilterWithNot.java]: > ... if the database fixture could not be set up.

And how would that happen anyway?

File: src/test/java/org/osiam/client/BugFilterWithNot.java [hunk 4]
```diff
@@ -0,0 +1,94 @@
...
+        SCIMSearchResult<Group> result = searchGroup(String.format("not(members eq \"%s\")", TEST_USER_ID));
+
+        assertThat(String.format("Can not find expected group %s!", NAME_GROUP_TWO),
+                result.getResources(), hasItem(new Group.Builder(NAME_GROUP_TWO).build()));
+    }
+
+    @Ignore("This test fails at the moment because the bug is not yet fixed. @Ignore must be removed after the bug-fix.")
+    @Test
+    public void filter_users_where_not_a_member_of_group() throws Exception {
+        Group expectedGroup = getGroupByName(NAME_GROUP_TWO);
```
Comment (tkrille) [inline src/test/java/org/osiam/client/BugFilterWithNot.java]: Why would you need to fetch the group from the database first? Just put the literal UUID of the group into the query (this makes the test faster, too).

**Comment by wallner** [comment]
Ok, it looks good to me now.

**Comment by wallner** [comment]
Thank you!

**Pull request merged**